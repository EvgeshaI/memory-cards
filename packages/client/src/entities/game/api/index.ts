export { sendLeaderToServer, fetchLeadersFromServer } from './leaderboardApi';
