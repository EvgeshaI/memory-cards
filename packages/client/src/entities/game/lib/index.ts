export { shuffleCards } from './shuffleCards';
export { checkMatch } from './checkMatch';
export { drawCards } from './drawCards';
export { handleEndGame } from './handleEndGame';
