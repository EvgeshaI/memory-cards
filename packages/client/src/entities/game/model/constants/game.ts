export const CARD_WIDTH = 100;
export const CARD_HEIGHT = 150;
export const GAP = 12;
