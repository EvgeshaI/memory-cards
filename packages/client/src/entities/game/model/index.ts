export * from './slice/gameSlice';
export * from './types';
export * from './selectors';
export * from '../lib';
