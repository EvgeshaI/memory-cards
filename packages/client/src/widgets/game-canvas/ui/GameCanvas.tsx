import React, { useEffect, useMemo, useRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from '@/shared/lib/store';
import {
  checkMatch,
  drawCards,
  gameActions,
  selectData,
  shuffleCards,
  handleGameTimeAndSubscription,
} from '@/entities/game';
import { RouteNames, routePaths } from '@/shared/constants/router';
import cls from './GameCanvas.module.scss';

export const GameCanvas = () => {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  const canvasRef = useRef<HTMLCanvasElement | null>(null);
  const cardSize = 100;
  const gap = 1;

  const { emojis: gameCards, numCards } = useAppSelector(selectData);

  const initialCards = useMemo(() => {
    const selectedEmojis = gameCards.slice(0, numCards / 2);
    return [...selectedEmojis, ...selectedEmojis];
  }, [gameCards, numCards]);

  const [time, setTime] = useState(0);
  const [cards, setCards] = useState<string[]>([]);
  const [openCards, setOpenCards] = useState<number[]>([]);
  const [matchedCards, setMatchedCards] = useState<number[]>([]);

  const cols = useMemo(() => Math.ceil(Math.sqrt(numCards)), [numCards]);

  useEffect(() => {
    const timerId = setInterval(() => {
      setTime((prevTime) => prevTime + 1);
    }, 1000);

    return () => clearInterval(timerId);
  }, []);

  useEffect(() => {
    setCards(shuffleCards([...initialCards]));
  }, [initialCards]);

  useEffect(() => {
    const canvas = canvasRef.current;
    if (canvas) {
      const ctx = canvas.getContext('2d');
      if (ctx) {
        drawCards(ctx, cards, openCards, matchedCards, cols, cardSize, gap);
      }
    }
  }, [cards, openCards, matchedCards, cols]);

  const handleClick = (event: React.MouseEvent<HTMLCanvasElement>) => {
    const rect = canvasRef.current!.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const y = event.clientY - rect.top;

    const cardIndex =
      Math.floor(y / (cardSize + gap)) * cols +
      Math.floor(x / (cardSize + gap));

    if (!openCards.includes(cardIndex) && !matchedCards.includes(cardIndex)) {
      setOpenCards((prev) => [...prev, cardIndex]);
      if (openCards.length === 1) {
        setTimeout(() => {
          checkMatch(
            openCards[0],
            cardIndex,
            cards,
            matchedCards,
            setMatchedCards,
            setOpenCards,
            async () => {
              dispatch(gameActions.saveGameTime(time));
              await handleGameTimeAndSubscription();

              navigate(routePaths[RouteNames.END_GAME]);
            },
          );
        }, 1000);
      }
    }
  };

  return (
    <canvas
      ref={canvasRef}
      className={cls.canvas}
      width={cols * (cardSize + gap)}
      height={Math.ceil(numCards / cols) * (cardSize + gap)}
      onClick={handleClick}
      data-testid="game-canvas"
    />
  );
};
