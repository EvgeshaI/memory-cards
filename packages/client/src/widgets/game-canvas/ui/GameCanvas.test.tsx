import { fireEvent, render } from '@testing-library/react'
import { act } from 'react'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import { MemoryRouter } from 'react-router-dom'
import { gameReducer } from '@/entities/game'
import { GameCanvas } from './GameCanvas'

interface GameState {
  numCards: number;
  emojis: string[];
  gameTime: number;
  leaders: []
}

const initialState: GameState = {
  numCards: 6,
  emojis: ['🎉', '😈', '🧠', '🐱', '🐶', '🍕', '🚀', '🌟', '🧁', '🍔'],
  gameTime: 0,
  leaders: []
};
jest.mock('@/entities/user', () => ({
  useUserData: jest.fn(() => ({
    user: {
      avatar: 'test-avatar-url',
      first_name: 'Test User',
      score: 100,
    },
  })),
}));

const createMockStore = (numCards: number) =>
  configureStore({
    reducer: { game: gameReducer },
    preloadedState: {
      game: {
        ...initialState,
        numCards,
      },
    },
  });

const renderWithStore = (store: ReturnType<typeof createMockStore>) =>
  render(
    <MemoryRouter>
      <Provider store={store}>
        <GameCanvas />
      </Provider>
    </MemoryRouter>,
  );

describe('GameCanvas - тесты со снимками', () => {
  test('рендер с 6 карточками', () => {
    const store = createMockStore(6);
    const { asFragment } = renderWithStore(store);

    expect(asFragment()).toMatchSnapshot();
  });

  test('рендер с 12 карточками', () => {
    const store = createMockStore(12);
    const { asFragment } = renderWithStore(store);

    expect(asFragment()).toMatchSnapshot();
  });

  test('рендер с 20 карточками', () => {
    const store = createMockStore(20);
    const { asFragment } = renderWithStore(store);

    expect(asFragment()).toMatchSnapshot();
  });

  test('состояние игрового поля после переворота карточки', () => {
    const store = createMockStore(6);
    const { asFragment, getByTestId } = renderWithStore(store);

    const canvas = getByTestId('game-canvas');
    act(() => {
      fireEvent.click(canvas, { clientX: 50, clientY: 50 });
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
