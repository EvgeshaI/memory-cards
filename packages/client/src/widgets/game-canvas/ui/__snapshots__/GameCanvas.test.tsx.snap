// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameCanvas - тесты со снимками рендер с 6 карточками 1`] = `
<DocumentFragment>
  <canvas
    data-testid="game-canvas"
    height="354"
    width="366"
  />
</DocumentFragment>
`;

exports[`GameCanvas - тесты со снимками рендер с 12 карточками 1`] = `
<DocumentFragment>
  <canvas
    data-testid="game-canvas"
    height="516"
    width="478"
  />
</DocumentFragment>
`;

exports[`GameCanvas - тесты со снимками рендер с 20 карточками 1`] = `
<DocumentFragment>
  <canvas
    data-testid="game-canvas"
    height="678"
    width="590"
  />
</DocumentFragment>
`;

exports[`GameCanvas - тесты со снимками состояние игрового поля после переворота карточки 1`] = `
<DocumentFragment>
  <canvas
    data-testid="game-canvas"
    height="354"
    width="366"
  />
</DocumentFragment>
`;
